unyt/array.py

```diff
@@ -1725,6 +1725,13 @@ def __getitem__(self, item):
             ret.units = self.units
         return ret
 
+    def __setitem__(self, item, value):
+        value_u = getattr(value, "units", None)
+        if value_u is not None:
+            if value_u != self.units:
+                value = value.to(self.units)
```

Comment:
mostly stylistic, although the behaviour is ever so slightly different (when both `self` and `value` have `None` for a `units` attribute, which of course should never happen)
```suggestion
        if hasattr(value, "units") and value.units != self.units:
            value = value.to(self.units)
```

Reply:
also, question: is this interoperable with `astropy.units` and `pint` ? (just curious)

Reply:
Astropy has `value.unit` instead of `value.units`, so that won't work. If I try to call `__setitem__` on a `unyt_array` with an astropy Quantity, I get a `TypeError` from astropy, but that's not new behavior with this PR.

For Pint I get a complicated traceback from Pint internals after calling `value.to(self.units)`.

I'm not overly upset over either situation.

Reply:
I have an irrational hatred of `hasattr` but you're right that's simpler.

Reply:
> I'm not overly upset over either situation.

me neither, just asking !

> I have an irrational hatred of hasattr

yeah, I get it. It can easily degenerate into terrible OOP.